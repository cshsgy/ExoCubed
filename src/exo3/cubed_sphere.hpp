#ifndef SRC_EXO3_CUBED_SPHERE_HPP_
#define SRC_EXO3_CUBED_SPHERE_HPP_

// C/C++
#include <memory>

// athena
#include <athena/athena.hpp>
#include <athena/mesh/mesh.hpp>

// canoe
#include <configure.hpp>

class CubedSphere {
 public:
  CubedSphere(MeshBlock *pmb);
  ~CubedSphere() {}

  static int FindBlockID(LogicalLocation const &loc);
  static void TransformOX(int *ox2, int *ox3, int *tox2, int *tox3,
                          LogicalLocation const &loc);

  static Real GenerateMeshX2(Real x, LogicalLocation const &loc);
  static Real GenerateMeshX3(Real x, LogicalLocation const &loc);

  void GetLatLon(Real *lat, Real *lon, int k, int j, int i) const;
  void GetLatLonFace2(Real *lat, Real *lon, int k, int j, int i) const;
  void GetLatLonFace3(Real *lat, Real *lon, int k, int j, int i) const;

  void GetUV(Real *U, Real *V, Real V2, Real V3, int k, int j, int i) const;
  void GetVyVz(Real *V2, Real *V3, Real U, Real V, int k, int j, int i) const;

  void TransformOX(int *ox2, int *ox3, int *tox2, int *tox3) const {
    return TransformOX(ox2, ox3, tox2, tox3, pmy_block_->loc);
  }

  void SaveLR3DValues(AthenaArray<Real> &L_in, AthenaArray<Real> &R_in,
                      int direction, int k, int j, int il, int iu);
  void LoadLR3DValues(AthenaArray<Real> &L_in, AthenaArray<Real> &R_in,
                      int direction, int k, int j, int il, int iu);

  void SynchronizeFluxesSend();
  void SynchronizeFluxesRecv();

 protected:
  void sendNeighborBlocks(int ox2, int ox3, int tg_rank, int tg_gid);
  void recvNeighborBlocks(int ox2, int ox3, int tg_rank, int tg_gid);

  AthenaArray<Real> L3DValues[3], R3DValues[3];

#ifdef MPI_PARALLEL
  MPI_Request send_request_[4];
  MPI_Request recv_request_[4];
#endif

  std::vector<Real> LRDataBuffer[4];

  MeshBlock *pmy_block_;
};

using CubedSpherePtr = std::shared_ptr<CubedSphere>;

#endif  // SRC_EXO3_CUBED_SPHERE_HPP_
