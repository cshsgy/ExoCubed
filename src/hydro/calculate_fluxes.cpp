//========================================================================================
// Athena++ astrophysical MHD code
// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
// Licensed under the 3-clause BSD License, see LICENSE file for details
//========================================================================================
//! \file calculate_fluxes.cpp
//! \brief Calculate hydro/MHD fluxes

// C headers

// C++ headers
#include <algorithm>   // min,max
#include <sstream>     // string stream

// Athena++ headers
#include <athena.hpp>
#include <athena_arrays.hpp>
#include <coordinates/coordinates.hpp>
#include <eos/eos.hpp>   // reapply floors to face-centered reconstructed states
#include <field/field.hpp>
#include <field/field_diffusion/field_diffusion.hpp>
#include <gravity/gravity.hpp>
#include <reconstruct/reconstruction.hpp>
#include <scalars/scalars.hpp>
#include <hydro/hydro.hpp>
#include <hydro/hydro_diffusion/hydro_diffusion.hpp>
#include <cubed_sphere.hpp>

// OpenMP header
#ifdef OPENMP_PARALLEL
#include <omp.h>
#endif

//----------------------------------------------------------------------------------------
//! \fn  void Hydro::CalculateFluxes
//! \brief Calculate Hydrodynamic Fluxes using the Riemann solver

void Hydro::CalculateFluxes(AthenaArray<Real> &w, FaceField &b,
                            AthenaArray<Real> &bcc, const int order) {
  MeshBlock *pmb = pmy_block;
  int is = pmb->is; int js = pmb->js; int ks = pmb->ks;
  int ie = pmb->ie; int je = pmb->je; int ke = pmb->ke;
  int il, iu, jl, ju, kl, ku;

  // b and bcc only used in Magnetic fields for now, should be able to do it quick
  // b,bcc are passed as fn parameters because clients may want to pass different bcc1,
  // b1, b2, etc., but the remaining members of the Field class are accessed directly via
  // pointers because they are unique. NOTE: b, bcc are nullptrs if no MHD.

  AthenaArray<Real> &flux_fc = scr1_nkji_;
  AthenaArray<Real> &laplacian_all_fc = scr2_nkji_;

  //--------------------------------------------------------------------------------------
  // i-direction

  AthenaArray<Real> &x1flux = flux[X1DIR];
  // set the loop limits

#ifndef HYDROSTATIC // sw not do X1

  jl = js, ju = je, kl = ks, ku = ke;

  for (int k=kl; k<=ku; ++k) {
    for (int j=jl; j<=ju; ++j) {
      // reconstruct L/R states
      if (order == 1) {
        pmb->precon->DonorCellX1(k, j, is-1, ie+1, w, bcc, wl_, wr_);
      } else if (order == 2) {
        pmb->precon->PiecewiseLinearX1(k, j, is-1, ie+1, w, bcc, wl_, wr_);
      } else {
        pmb->precon->PiecewiseParabolicX1(k, j, is-1, ie+1, w, bcc, wl_, wr_);
      }
      pmb->pcoord->CenterWidth1(k, j, is, ie+1, dxw_);
#ifdef CUBED_SPHERE // Rieman solver run later
      SaveLR3DValues(wl_, wr_, X1DIR, k, j, is, ie+1); // is to ie+1 is what the RiemannSolver below uses...
#else
      RiemannSolver(k, j, is, ie+1, IVX, wl_, wr_, x1flux, dxw_);

#endif
    }
  }
#endif  // end HYDROSTATIC

  //--------------------------------------------------------------------------------------
  // j-direction
  if (pmb->pmy_mesh->f2) {
    AthenaArray<Real> &x2flux = flux[X2DIR];
    // set the loop limits
    il = is-1, iu = ie+1, kl = ks, ku = ke;

    for (int k=kl; k<=ku; ++k) {
      // reconstruct the first row
      if (order == 1) {
        pmb->precon->DonorCellX2(k, js-1, il, iu, w, bcc, wl_, wr_);
      } else if (order == 2) {
        pmb->precon->PiecewiseLinearX2(k, js-1, il, iu, w, bcc, wl_, wr_);
      } else {
        pmb->precon->PiecewiseParabolicX2(k, js-1, il, iu, w, bcc, wl_, wr_);
      }
      for (int j=js; j<=je+1; ++j) {
        // reconstruct L/R states at j
        if (order == 1) {
          pmb->precon->DonorCellX2(k, j, il, iu, w, bcc, wlb_, wr_);
        } else if (order == 2) {
          pmb->precon->PiecewiseLinearX2(k, j, il, iu, w, bcc, wlb_, wr_);
        } else {
          pmb->precon->PiecewiseParabolicX2(k, j, il, iu, w, bcc, wlb_, wr_);
        }
        pmb->pcoord->CenterWidth2(k, j, il, iu, dxw_);
#ifdef CUBED_SPHERE // Rieman solver run later
        SaveLR3DValues(wl_, wr_, X2DIR, k, j, il, iu); // il to iu is what the RiemannSolver below uses...
        wl_.SwapAthenaArray(wlb_);
#else
        RiemannSolver(k, j, il, iu, IVY, wl_, wr_, x2flux, dxw_);
        // swap the arrays for the next step
        wl_.SwapAthenaArray(wlb_);

#endif
      }
    }
  }

  //--------------------------------------------------------------------------------------
  // k-direction
  if (pmb->pmy_mesh->f3) {
    AthenaArray<Real> &x3flux = flux[X3DIR];
    // set the loop limits
    il = is, iu = ie, jl = js, ju = je;

    for (int j=jl; j<=ju; ++j) { // this loop ordering is intentional
      // reconstruct the first row
      if (order == 1) {
        pmb->precon->DonorCellX3(ks-1, j, il, iu, w, bcc, wl_, wr_);
      } else if (order == 2) {
        pmb->precon->PiecewiseLinearX3(ks-1, j, il, iu, w, bcc, wl_, wr_);
      } else {
        pmb->precon->PiecewiseParabolicX3(ks-1, j, il, iu, w, bcc, wl_, wr_);
      }
      for (int k=ks; k<=ke+1; ++k) {
        // reconstruct L/R states at k
        if (order == 1) {
          pmb->precon->DonorCellX3(k, j, il, iu, w, bcc, wlb_, wr_);
        } else if (order == 2) {
          pmb->precon->PiecewiseLinearX3(k, j, il, iu, w, bcc, wlb_, wr_);
        } else {
          pmb->precon->PiecewiseParabolicX3(k, j, il, iu, w, bcc, wlb_, wr_);
        }
        pmb->pcoord->CenterWidth3(k, j, il, iu, dxw_);
#ifdef CUBED_SPHERE // Rieman solver run later
        SaveLR3DValues(wl_, wr_, X3DIR, k, j, il, iu); // il to iu is what the RiemannSolver below uses...
        wl_.SwapAthenaArray(wlb_);
#else

        RiemannSolver(k, j, il, iu, IVZ, wl_, wr_, x3flux, dxw_);
        // swap the arrays for the next step
        wl_.SwapAthenaArray(wlb_);

#endif
      }
    }
  }

// Cubed Sphere: recover the stored values, run riemann solvers
#ifdef CUBED_SPHERE
  // Temporarily comment out the following 2 lines to avoid the error
  SynchronizeFluxesSend();
  SynchronizeFluxesRecv();
  //--------------------------------------------------------------------------------------
  // i-direction
#ifndef HYDROSTATIC
  jl = js, ju = je, kl = ks, ku = ke;

  for (int k=kl; k<=ku; ++k) {
    for (int j=jl; j<=ju; ++j) {
      // reconstruct L/R states
      LoadLR3DValues(wl_, wr_, X1DIR, k, j, is, ie+1);
      pmb->pcoord->CenterWidth1(k, j, is, ie+1, dxw_);
      RiemannSolver(k, j, is, ie+1, IVX, wl_, wr_, x1flux, dxw_);
    }
  }
#endif // HYDROSTATIC

  //--------------------------------------------------------------------------------------
  // j-direction

  if (pmb->pmy_mesh->f2) {
    AthenaArray<Real> &x2flux = flux[X2DIR];
    // set the loop limits
    il = is-1, iu = ie+1, kl = ks, ku = ke;

    for (int k=kl; k<=ku; ++k) {
      for (int j=js; j<=je+1; ++j) {
        // reconstruct L/R states at j
        pmb->pcoord->CenterWidth2(k, j, il, iu, dxw_);
        LoadLR3DValues(wl_, wr_, X2DIR, k, j, il, iu); // il to iu is what the RiemannSolver below uses...
        RiemannSolver(k, j, il, iu, IVY, wl_, wr_, x2flux, dxw_);
      }
    }
  }


  //--------------------------------------------------------------------------------------
  // k-direction

  if (pmb->pmy_mesh->f3) {
    AthenaArray<Real> &x3flux = flux[X3DIR];
    // set the loop limits
    il = is, iu = ie, jl = js, ju = je;

    for (int j=jl; j<=ju; ++j) { // this loop ordering is intentional
      for (int k=ks; k<=ke+1; ++k) {
        pmb->pcoord->CenterWidth3(k, j, il, iu, dxw_);
        LoadLR3DValues(wl_, wr_, X3DIR, k, j, il, iu);

        RiemannSolver(k, j, il, iu, IVZ, wl_, wr_, x3flux, dxw_);
      }
    }
  }
#endif

  if (!STS_ENABLED)
    AddDiffusionFluxes();

  return;
}

//----------------------------------------------------------------------------------------
//! \fn  void Hydro::CalculateFluxes_STS
//! \brief Calculate Hydrodynamic Diffusion Fluxes for STS

void Hydro::CalculateFluxes_STS() {
  AddDiffusionFluxes();
}

void Hydro::AddDiffusionFluxes() {
  Field *pf = pmy_block->pfield;
  // add diffusion fluxes
  if (hdif.hydro_diffusion_defined) {
    if (hdif.nu_iso > 0.0 || hdif.nu_aniso > 0.0)
      hdif.AddDiffusionFlux(hdif.visflx,flux);
    if (NON_BAROTROPIC_EOS) {
      if (hdif.kappa_iso > 0.0 || hdif.kappa_aniso > 0.0)
        hdif.AddDiffusionEnergyFlux(hdif.cndflx,flux);
    }
  }
  if (MAGNETIC_FIELDS_ENABLED && NON_BAROTROPIC_EOS) {
    if (pf->fdif.field_diffusion_defined)
      pf->fdif.AddPoyntingFlux(pf->fdif.pflux);
  }
  return;
}
