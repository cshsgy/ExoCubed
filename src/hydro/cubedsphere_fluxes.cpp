#include <sstream>
#include <iostream>

#include <globals.hpp>
#include <hydro/hydro.hpp>
#include <cubed_sphere.hpp>

#ifdef MPI_PARALLEL
  #include <mpi.h>
#endif

void Hydro::SaveLR3DValues(AthenaArray<Real> &L_in, AthenaArray<Real> &R_in,
  int direction,  int k, int j, int il, int iu) {
    for (int n=0; n<NWAVE; n++){
        for (int i=il; i<=iu; i++){
            L3DValues[direction](n, k, j, i) = L_in(n, i);
            R3DValues[direction](n, k, j, i) = R_in(n, i);
        }
    }
}

void Hydro::LoadLR3DValues(AthenaArray<Real> &L_in, AthenaArray<Real> &R_in,
  int direction,  int k, int j, int il, int iu) {
    for (int n=0; n<NWAVE; n++){
        for (int i=il; i<=iu; i++){
            L_in(n, i) = L3DValues[direction](n, k, j, i);
            R_in(n, i) = R3DValues[direction](n, k, j, i);
        }
    }
}

void Hydro::SynchronizeFluxesSend(){
    MeshBlock *pmb = pmy_block;
    for (int n=0; n<pmb->pbval->nneighbor; n++){
        NeighborBlock &nb = pmb->pbval->neighbor[n];
        if(nb.ni.ox1==0 && nb.ni.ox2*nb.ni.ox3==0){ // On x2 and x3 face boundaries only
            SendNeighborBlocks(pmb->loc, nb.ni.ox2, nb.ni.ox3, nb.snb.rank, nb.snb.gid);
        }
    }
}

void Hydro::SynchronizeFluxesRecv(){
    MeshBlock *pmb = pmy_block;
    for (int n=0; n<pmb->pbval->nneighbor; n++){
        NeighborBlock &nb = pmb->pbval->neighbor[n];
        if(nb.ni.ox1==0 && nb.ni.ox2*nb.ni.ox3==0){ // On x2 and x3 face boundaries only
            RecvNeighborBlocks(pmb->loc, nb.ni.ox2, nb.ni.ox3, nb.snb.rank, nb.snb.gid);
        }
    }
}

void Hydro::SendNeighborBlocks(LogicalLocation const& loc, int ox2, int ox3, int tg_rank, int tg_gid){
    MeshBlock *pmb = pmy_block;

    int lv2_lx2 = loc.lx2 >> (loc.level - 2);
    int lv2_lx3 = loc.lx3 >> (loc.level - 2);
    int blockID = lv2_lx2 + lv2_lx3 * 2 + 1;
    // Calculate local ID
    int local_lx2 = loc.lx2 - (lv2_lx2<<(loc.level - 2));
    int local_lx3 = loc.lx3 - (lv2_lx3<<(loc.level - 2));
    int bound_lim = (1<<(loc.level - 2)) - 1;
    int tox2, tox3;
    int DirTag, DirNum, ownTag; // Tag for target, and the numbering of axis
    int ox2_bkp = ox2;
    int ox3_bkp = ox3;
    bool invDir, Left; // Marking whether need to reverse direction in packing or unpacking; Left marking left or right.
    // Hard code the boundary cases
    if(local_lx2==bound_lim && ox2==1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = true;
        // Determine whether need to inverse direction
        switch(blockID){
            case 1: invDir = false; break;
            case 2: invDir = false; break;
            case 3: invDir = true; break;
            case 4: invDir = false; break;
            case 5: invDir = true; break;
            case 6: invDir = true; break;
        }
    }else if(local_lx2==0 && ox2==-1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = false;
        switch(blockID){
            case 1: invDir = true; break;
            case 2: invDir = false; break;
            case 3: invDir = false; break;
            case 4: invDir = true; break;
            case 5: invDir = false; break;
            case 6: invDir = true; break;
        }
    }else if(local_lx3==bound_lim && ox3==1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = true;
        switch(blockID){
            case 1: invDir = true; break;
            case 2: invDir = false; break;
            case 3: invDir = false; break;
            case 4: invDir = false; break;
            case 5: invDir = false; break;
            case 6: invDir = false; break;
        }
    }else if(local_lx3==0 && ox3==-1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = false;
        switch(blockID){
            case 1: invDir = false; break;
            case 2: invDir = false; break;
            case 3: invDir = false; break;
            case 4: invDir = false; break;
            case 5: invDir = true; break;
            case 6: invDir = false; break;
        }
    }else{ // No need to communicate fluxes, return
        return;
    }
    // Pack the data
    int kb1, kb2, jb1, jb2, ib1, ib2;
    if (ox2==1){
        DirNum = X2DIR;
        jb1 = pmb->je + 1;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ke + 1;
    }
    if (ox2==-1){
        DirNum = X2DIR;
        jb1 = pmb->js;
        jb2 = pmb->js;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ke + 1;
    }
    if (ox3==1){
        DirNum = X3DIR;
        jb1 = pmb->js;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ke + 1;
        kb2 = pmb->ke + 1;
    }
    if (ox3==-1){
        DirNum = X3DIR;
        jb1 = pmb->js;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ks;
    }
    int dsize = ((kb2 - kb1 + 1) * (jb2 - jb1 + 1) * (ib2 - ib1 + 1) * NWAVE);
    Real *data = new Real[dsize];
    int offset = 0;
    if (invDir){
        for (int n=0; n<NWAVE; n++)
            for (int k=kb2; k>=kb1; k--)
                for (int j=jb2; j>=jb1; j--)
                    for (int i=ib1; i<=ib2; i++)
                        if (Left)
                            data[offset++] = L3DValues[DirNum](n,k,j,i);
                        else
                            data[offset++] = R3DValues[DirNum](k,j,i);
    }else{
        for (int n=0; n<NWAVE; n++)
            for (int k=kb1; k<=kb2; k++)
                for (int j=jb1; j<=jb2; j++)
                    for (int i=ib1; i<=ib2; i++)
                        if (Left)
                            data[offset++] = L3DValues[DirNum](n,k,j,i);
                        else
                            data[offset++] = R3DValues[DirNum](n,k,j,i);
    }

    // Calculate the tag of destination
    if (tox2==-1) DirTag = 0 + 4 * pmb->gid + 24*(1 << (loc.level - 2)) * tg_gid;
    if (tox2==1) DirTag = 1 + 4 * pmb->gid + 24*(1 << (loc.level - 2)) * tg_gid;
    if (tox3==-1) DirTag = 2 + 4 * pmb->gid + 24*(1 << (loc.level - 2)) * tg_gid;
    if (tox3==1) DirTag = 3 + 4 * pmb->gid + 24*(1 << (loc.level - 2)) * tg_gid;
    // Send by MPI: we don't care whether it is in the same process for now
    if (ox2==-1) ownTag = 0;
    if (ox2==1) ownTag = 1;
    if (ox3==-1) ownTag = 2;
    if (ox3==1) ownTag = 3;
    MPI_Isend(data, dsize, MPI_DOUBLE, tg_rank, DirTag, MPI_COMM_WORLD, &send_request[ownTag]);
    std::cout << "===============================" << std::endl;
    std::cout << "MPI Message: Sent data with size " << dsize << " from rank " << Globals::my_rank << " to " << tg_rank << " on tag number " << DirTag << std::endl;
    std::cout << "This is a message from block gid " << pmb->gid << " to " << tg_gid << std::endl;
    std::cout << "The direction is ox2=" << ox2 << ", ox3=" << ox3 << " and inversion is " << invDir << std::endl;
    delete[] data;
}

void Hydro::RecvNeighborBlocks(LogicalLocation const& loc, int ox2, int ox3, int tg_rank, int tg_gid){
    MeshBlock *pmb = pmy_block;

    int lv2_lx2 = loc.lx2 >> (loc.level - 2);
    int lv2_lx3 = loc.lx3 >> (loc.level - 2);
    int blockID = lv2_lx2 + lv2_lx3 * 2 + 1;
    // Calculate local ID
    int local_lx2 = loc.lx2 - (lv2_lx2<<(loc.level - 2));
    int local_lx3 = loc.lx3 - (lv2_lx3<<(loc.level - 2));
    int bound_lim = (1<<(loc.level - 2)) - 1;
    int tox2, tox3;
    int DirTag, DirNum, ownTag; // Tag for receiving, and the numbering of axis to place the values
    int ox2_bkp = ox2;
    int ox3_bkp = ox3;
    bool invDir, Left; // Marking whether need to reverse direction in packing or unpacking; Left marking left or right.
    // Hard code the boundary cases
    if(local_lx2==bound_lim && ox2==1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = false;
        // Determine whether need to inverse direction
    }else if(local_lx2==0 && ox2==-1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = true;
    }else if(local_lx3==bound_lim && ox3==1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = false;
    }else if(local_lx3==0 && ox3==-1){
        TransformOxForCubedSphere(&ox2_bkp, &ox3_bkp, &tox2, &tox3, loc);
        Left = true;
    }else{ // No need to communicate fluxes, return
        return;
    }
    // Pack the data
    int kb1, kb2, jb1, jb2, ib1, ib2;
    if (ox2==1){
        DirNum = X2DIR;
        jb1 = pmb->je + 1;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ke + 1;
    }
    if (ox2==-1){
        DirNum = X2DIR;
        jb1 = pmb->js;
        jb2 = pmb->js;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ke + 1;
    }
    if (ox3==1){
        DirNum = X3DIR;
        jb1 = pmb->js;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ke + 1;
        kb2 = pmb->ke + 1;
    }
    if (ox3==-1){
        DirNum = X3DIR;
        jb1 = pmb->js;
        jb2 = pmb->je + 1;
        ib1 = pmb->is;
        ib2 = pmb->ie + 1;
        kb1 = pmb->ks;
        kb2 = pmb->ks;
    }
    int dsize = ((kb2 - kb1 + 1) * (jb2 - jb1 + 1) * (ib2 - ib1 + 1) * NWAVE);
    Real *data = new Real[dsize];
    // Calculate the tag for receiving
    if (ox2==-1) DirTag = 0 + 4 * tg_gid + 24*(1 << (loc.level - 2)) * pmb->gid;
    if (ox2==1) DirTag = 1 + 4 * tg_gid + 24*(1 << (loc.level - 2)) * pmb->gid;
    if (ox3==-1) DirTag = 2 + 4 * tg_gid + 24*(1 << (loc.level - 2)) * pmb->gid;
    if (ox3==1) DirTag = 3 + 4 * tg_gid + 24*(1 << (loc.level - 2)) * pmb->gid;
    MPI_Recv(data, dsize, MPI_DOUBLE, tg_rank, DirTag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);    
    std::cout << "===============================" << std::endl;
    std::cout << "MPI Message: Received data from " << tg_gid << " to " << pmb->gid << " on tag number " << DirTag << std::endl;

    // =======
    // int test;
    // probe MPI communications.  This is a bit of black magic that seems to promote
    // communications to top of stack and gets them to complete more quickly
    // MPI_Iprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &test,
    //             MPI_STATUS_IGNORE);
    // =======
    // MPI_Test(&recv_request[ownTag], &test, MPI_STATUS_IGNORE);
    

    int offset = 0;
    for (int n=0; n<NWAVE; n++)
        for (int k=kb1; k<=kb2; k++)
            for (int j=jb1; j<=jb2; j++)
                for (int i=ib1; i<=ib2; i++)
                    if (Left)
                        L3DValues[DirNum](n,k,j,i) = data[offset++];
                    else
                        R3DValues[DirNum](n,k,j,i) = data[offset++];
    delete[] data;
}
