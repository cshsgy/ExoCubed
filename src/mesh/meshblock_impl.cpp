// Athena++ headers
#include <athena.hpp>
#include <parameter_input.hpp>
#include <coordinates/coordinates.hpp>
#include <hydro/hydro.hpp>
#include <field/field.hpp>
#include <scalars/scalars.hpp>
#include <orbital_advection/orbital_advection.hpp>
#include <gravity/gravity.hpp>
#include <eos/eos.hpp>

// exdacs headers
#include <configure.hpp>
#include "block_index.hpp"
#include "meshblock_impl.hpp"
#include "../hydro/implicit/implicit_solver.hpp"
#include "../hydro/decomposition/decomposition.hpp"
#include "../reconstruct/face_reconstruct.hpp"

MeshBlock::Impl::Impl(MeshBlock *pmb, ParameterInput *pin):
  pmy_block_(pmb)
{
  du.NewAthenaArray(NumHydros, pmb->ncells3, pmb->ncells2, pmb->ncells1);

  // block index
  pblock = new BlockIndex(pmb);

  // hydro decomposition
  pdec = new Decomposition(pmb->phydro);

  // implicit methods
  phevi = new ImplicitSolver(pmb, pin);

  // reconstruction
  precon = new FaceReconstruct(pmb, pin);
}

// Athena++ demands destruct pbval AFTER all boundary values
// But in this mod, boundary values are destructed BEFORE pbval
// TODO, check if this is OK
MeshBlock::Impl::~Impl()
{
  //std::cout << "Impl desctructor" << std::endl;
  delete pblock;
  delete pdec;
  delete phevi;
  delete precon;
}
