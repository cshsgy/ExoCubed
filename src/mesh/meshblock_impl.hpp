#ifndef MESHBLOCK_IMPL_HPP
#define MESHBLOCK_IMPL_HPP

// C/C++ header
#include <memory>

// Athena++ header
#include <mesh/mesh.hpp>

class BlockIndex;
class ParameterInput;
class Decomposition;
class ImplicitSolver;
class FaceReconstruct;
class Forcing;

//! \class CellVariables
//  \brief a collection of all physical data in a computational cell
class MeshBlock::Impl {
public:
  Impl(MeshBlock *pmb, ParameterInput *pin);
  ~Impl();

  AthenaArray<Real> du;    // stores tendency

  BlockIndex            *pblock;
  Decomposition         *pdec;
  ImplicitSolver        *phevi;
  FaceReconstruct       *precon;
  Forcing               *pforce;

private:
  MeshBlock const *pmy_block_;
};

#endif
