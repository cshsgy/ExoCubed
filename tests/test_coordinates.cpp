#include <parameter_input.hpp>
#include <hydro/hydro.hpp>
#include <mesh/mesh.hpp>
#include <configure.hpp>
#include <cubed_sphere.hpp>

void MeshBlock::ProblemGenerator(ParameterInput *pin)
{
  std::cout << "my rank  = " << Globals::my_rank << std::endl;
  std::cout << "my panel = " << FindBlockID(loc) << std::endl;

  std::cout << "x1v = " << std::endl;
  for (int i = is - GhostZoneSize; i < is; ++i) {
    std::cout << pcoord->x1v(i)  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;

  for (int i = is; i <= ie; ++i) {
    std::cout << pcoord->x1v(i)  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;
  for (int i = ie + 1; i <= ie + GhostZoneSize; ++i) {
    std::cout << pcoord->x1v(i)  << " " << std::endl;
  }
  std::cout << std::endl;

  std::cout << "x2f = " << std::endl;
  for (int j = js - GhostZoneSize; j < js; ++j) {
    std::cout << pcoord->x2f(j)/M_PI  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;

  for (int j = js; j <= je; ++j) {
    std::cout << pcoord->x2f(j)/M_PI  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;

  for (int j = je + 1; j <= je + GhostZoneSize; ++j) {
    std::cout << pcoord->x2f(j)/M_PI  << " " << std::endl;
  }
  std::cout << std::endl;

  std::cout << "x3f = " << std::endl;
  for (int k = ks - GhostZoneSize; k < ks; ++k) {
    std::cout << pcoord->x3f(k)/M_PI  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;

  for (int k = ks; k <= ke; ++k) {
    std::cout << pcoord->x3f(k)/M_PI  << " " << std::endl;
  }
  std::cout << "--------------" << std::endl;

  for (int k = ke + 1; k <= ke + GhostZoneSize; ++k) {
    std::cout << pcoord->x3f(k)/M_PI  << " " << std::endl;
  }
}
