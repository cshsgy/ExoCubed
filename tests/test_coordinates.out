
######################################################
##                 EXDACS MODEL                     ##
######################################################
1. Initializing Main ...
my rank  = 0
my panel = 1
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 
my rank  = 0
my panel = 2
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 
my rank  = 0
my panel = 3
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 
my rank  = 0
my panel = 4
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 
my rank  = 0
my panel = 5
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
-2.77556e-17 
0.125 
--------------
0.25 
0.375 
my rank  = 0
my panel = 6
x1v = 
-0.375 
-0.125 
--------------
0.125 
0.375 
0.625 
0.875 
--------------
1.125 
1.375 

x2f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
0 
0.125 
--------------
0.25 
0.375 

x3f = 
-0.5 
-0.375 
--------------
-0.25 
-0.125 
-2.77556e-17 
0.125 
--------------
0.25 
0.375 

Setup complete, entering main loop...

cycle=0 time=0.0000000000000000e+00 dt=9.0000000000000000e+01
cycle=1 time=9.0000000000000000e+01 dt=1.8000000000000000e+02
cycle=1 time=9.0000000000000000e+01 dt=1.8000000000000000e+02

Terminating on cycle limit
time=9.0000000000000000e+01 cycle=1
tlim=9.0000000000000000e+01 nlim=1

zone-cycles = 384
cpu time used  = 4.5367999999999999e-02
zone-cycles/cpu_second = 8.4641156762475748e+03
Main done.
